# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Data types

In [2]:
path = r'C:\Users\Kozole\Desktop\CF Kurs\Excercise\4\Instacart Basket Analysis'

In [3]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_4.8.pkl'))

In [4]:
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [5]:
df_ords_prods_merge.shape

(32406041, 24)

In [6]:
df_cust.shape

(206209, 10)

# 03. Task 4.9 Part 1

In [8]:
# 04.Wrangle the data so that it follows consistent logic; for example, 
# rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [10]:
# Checking column names

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [11]:
# Checking for null values
df_cust.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

# There are 11259 missing values in the First Name column. I have decided to drop First Name and Surnam columns because I think they don’t do anything for analysis. We can always retrieve this information from the original file and link it to the user_id column.

In [12]:
# Dropping First Name and Surnam columns

df_cust=df_cust.drop(columns=['First Name','Surnam'])

In [13]:
# Checking if columns are dropped

df_cust.head()

user_id  Gender       STATE  Age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [14]:
# Renaming columns


df_cust.rename(columns={'Gender':'gender','STATE':'state','Age':'age','n_dependants':'number_of_dependants','fam_status':'family_status'},inplace=True)

In [15]:
# Checking if columns are renamed

df_cust.head()

user_id  gender       state  age date_joined  number_of_dependants  \
0    26711  Female    Missouri   48    1/1/2017                     3   
1    33890  Female  New Mexico   36    1/1/2017                     0   
2    65803    Male       Idaho   35    1/1/2017                     2   
3   125935  Female        Iowa   40    1/1/2017                     0   
4   130797  Female    Maryland   26    1/1/2017                     1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [16]:
# 05. Complete the fundamental data quality and consistency checks you’ve 
# learned throughout this Achievement; for example, check for and address 
# missing values and duplicates, and convert any mixed-type data.

In [17]:
# Checking for missing values
df_cust.isnull().sum()

user_id                 0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependants    0
family_status           0
income                  0
dtype: int64

# There are no missing values in the dataframe

In [18]:
# Checking for duplicates

df_cust[df_cust.duplicated()]

Empty DataFrame
Columns: [user_id, gender, state, age, date_joined, number_of_dependants, family_status, income]
Index: []

# There are no duplicates in the dataframe

In [19]:
# Checking for mixed-type data

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

# There are no mixed-type data in the dataframe

In [20]:
# 06. Combine your customer data with the rest of your prepared Instacart data. 
# (Hint: Make sure the key columns are the same data type!)

In [21]:
df_cust.shape

(206209, 8)

In [22]:
# Checking columns data types od the df_cust dataframe

df_cust.dtypes

user_id                  int64
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
family_status           object
income                   int64
dtype: object

In [24]:
# Checking columns data types od the df_ords_prods_merge dataframe

df_ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_price              float64
spenders_flag               object
frequency_median           float64
frequency_flag              object
dtype: object

# The key column for combining is user_id and it should be changed to String because it does not represent real numeric values, rather, keys related to specific users

In [25]:
# Changing column user_id data type for df_cust dataframe

df_cust['user_id'] = df_cust['user_id'].astype('str')

In [26]:
# Checking columns data types od the df_cust dataframe

df_cust.dtypes

user_id                 object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
family_status           object
income                   int64
dtype: object

In [27]:
# Changing column user_id data type for df_ords_prods_merge dataframe

df_ords_prods_merge['user_id'] = df_ords_prods_merge['user_id'].astype('str')

In [28]:
# Checking columns data types od the df_ords_prods_merge dataframe

df_ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                     object
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_price              float64
spenders_flag               object
frequency_median           float64
frequency_flag              object
dtype: object

# Both key columns are the same data type

In [29]:
# Merging the data sets on column user_id

df_ords_prods_all=df_ords_prods_merge.merge(df_cust,on='user_id',indicator='True')

In [30]:
df_ords_prods_all.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2         907        Premium Sliced Bacon       106             12    20.0   
3         907        Premium Sliced Bacon       106             12    20.0   
4        1000                    Apricots        18             10    12.9   

   order_id user_id  order_number  orders_day_of_week  order_hour_of_day  ...  \
0   3139998     138            28                   6                 11  ...   
1   1977647     138            30                   6                 17  ...   
2   3160996     138             1                   5                 13  ...   
3   2254091     138            10                   5                 14  ...   
4    505689     138             9                   6                 12  ...   

   frequency_median     frequency_flag  gender      state age date_joined  \
0               8.0  Frequent customer    Male  Minnesota  81    8/1/2019   
1               8.0  Frequent customer    Male  Minnesota  81    8/1/2019   
2               8.0  Frequent customer    Male  Minnesota  81    8/1/2019   
3               8.0  Frequent customer    Male  Minnesota  81    8/1/2019   
4               8.0  Frequent customer    Male  Minnesota  81    8/1/2019   

  number_of_dependants family_status  income  True  
0                    1       married   49620  both  
1                    1       married   49620  both  
2                    1       married   49620  both  
3                    1       married   49620  both  
4                    1       married   49620  both  

[5 rows x 32 columns]

In [31]:
# Checking the results of the merge

df_ords_prods_all['True'].value_counts()

both          32406041
left_only            0
right_only           0
Name: True, dtype: int64

In [32]:
# Checking the results of the merge

df_ords_prods_all.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                     object
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_price              float64
spenders_flag               object
frequency_median           float64
frequency_flag              object
gender                      object
state                       object
age                          int64
date_joined                 object
number_of_dependants

In [33]:
# Dropping the True column created by the merge

df_ords_prods_all=df_ords_prods_all.drop(columns='True')

In [34]:
# Checking the results of the drop

df_ords_prods_all.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                     object
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_price              float64
spenders_flag               object
frequency_median           float64
frequency_flag              object
gender                      object
state                       object
age                          int64
date_joined                 object
number_of_dependants

In [35]:
# Exporting the dataframe

df_ords_prods_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all_4.9.pkl'))

In [36]:
df_ords_prods_all.shape

(32406041, 31)